File: PhotoViewer-ObjC/AlbumViewController.m
//
//  AlbumViewController.m
//  PhotoViewer
//
//  Created by ASIM27 on 1/20/17.
//  Copyright © 2017 km3h. All rights reserved.
//

#import "AlbumViewController.h"
#import "PhotosViewController.h"
#import "AlbumTableViewCell.h"

@interface AlbumViewController ()
@property (strong, nonatomic) PhotosViewController *photosViewController;
@property (strong, nonatomic) NSArray *albumIds;
@end

@implementation AlbumViewController

-(instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self.splitViewController action:NSSelectorFromString(@"backButtonTapped")];
    self.navigationItem.leftBarButtonItem = backButtonItem;
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    NSArray *controllers = self.splitViewController.viewControllers;
    self.photosViewController = (PhotosViewController*)((UINavigationController*)[controllers lastObject]).topViewController;
}

-(void)viewWillAppear:(BOOL)animated
{
    self.clearsSelectionOnViewWillAppear = self.splitViewController.isCollapsed;
    [self reloadData];
    [super viewWillAppear:animated];
}

-(void)reloadData
{
    [self.dataSource dataHandler:^(NSDictionary *json) {
        dispatch_sync(dispatch_get_main_queue(), ^{
            NSArray *keys = [json allKeys];
            NSArray *sortedKeys = [keys sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull n1, id  _Nonnull n2) { return [n1 intValue] > [n2 intValue];}];
            self.albumIds = sortedKeys;
            [self.tableView reloadData];
        });
    }];
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqual:@"showDetail"])
    {
        NSIndexPath *indexPath = self.tableView.indexPathForSelectedRow;
        PhotosViewController *controller = (PhotosViewController*)((UINavigationController*)(segue.destinationViewController)).topViewController;
        NSNumber *albumId = self.albumIds[indexPath.row];
        controller.albumId = albumId;
        controller.dataSource = self.dataSource;
        controller.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
        controller.navigationItem.leftItemsSupplementBackButton = YES;
    }
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.albumIds.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    AlbumTableViewCell *cell = (AlbumTableViewCell*)[tableView dequeueReusableCellWithIdentifier:@"Cell"];
    
    NSNumber *object = self.albumIds[indexPath.row];
    cell.sideLabel.text = [NSString stringWithFormat:@"Album %@", object.description];
    return cell;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 200.0;
}

@end

File: PhotoViewer/AlbumViewController.swift
//
//  AlbumViewController.swift
//  PhotoViewer
//
//  Created by ASIM27 on 11/16/16.
//  Copyright © 2016 km3h. All rights reserved.
//

import UIKit

class AlbumViewController: UITableViewController
{
    public weak var dataSource : DataSource?
    
    var photosViewController: PhotosViewController? = nil
    var albumIds : [NSNumber]?
    
    required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
        let selector = Selector(("backButtonTapped"))
        let backButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self.splitViewController, action: selector)
        self.navigationItem.leftBarButtonItem = backButtonItem
    }
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        if let split = self.splitViewController
        {
            let controllers = split.viewControllers
            self.photosViewController = (controllers[controllers.count-1] as! UINavigationController).topViewController as? PhotosViewController
        }
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        self.clearsSelectionOnViewWillAppear = self.splitViewController!.isCollapsed
        self.reloadData()
        super.viewWillAppear(animated)
    }
    
    public func reloadData()
    {
        self.dataSource?.photoDictionary() { json in
            DispatchQueue.main.async() {
                let keys = json?.keys
                let sortedKeys = keys?.sorted() { n1, n2 in
                    n1.intValue < n2.intValue
                }
                self.albumIds = sortedKeys
                self.tableView.reloadData()
            }
        }
    }
    
    // MARK: - Segues
    override func prepare(for segue: UIStoryboardSegue, sender: Any?)
    {
        if segue.identifier == "showDetail" {
            if let indexPath = self.tableView.indexPathForSelectedRow {
                let controller = (segue.destination as! UINavigationController).topViewController as! PhotosViewController
                if let albumId = self.albumIds?[indexPath.row] {
                    controller.albumId = albumId
                }
                controller.dataSource = self.dataSource
                controller.navigationItem.leftBarButtonItem = self.splitViewController?.displayModeButtonItem
                controller.navigationItem.leftItemsSupplementBackButton = true
            }
        }
    }
    
    // MARK: - table view delegate methods
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        if let albumIds = self.albumIds
        {
            return albumIds.count
        }
        else
        {
            return 0
        }
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath) as! AlbumTableViewCell
        
        if let object = self.albumIds?[indexPath.row]
        {
            cell.sideLabel!.text = ("Album \(object.description)")
        }
        return cell
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat
    {
        return 200.0
    }
}

